Mathlib/Tactic/Simproc/ExistsAndEqNested.lean

```diff
@@ -0,0 +1,430 @@
+/-
+Copyright (c) 2025 Vasilii Nesterov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Vasilii Nesterov
+-/
+import Mathlib.Init
+import Qq
+
+/-!
+# simproc for `∃ a', ... ∧ a' = a ∧ ...`
+
+This module implements the `existsAndEq` simproc that triggers on goals of the form `∃ a, P`,
+checks if `P` allows only one value for `a` and if so, substitute it elimination leading quantifier.
+It traverses the body branching at each `∧` and enterring existential quantifiers looking for a
+subexpression of the form `a = a'` or `a' = a`. If such expression found all occurances of `a`
+are replaced with `a'`. If `a'` depends on some variables bounded by enterred existential
+quantifiers those quantifiers are moved outside.
+
+For example `∃ a, p a ∧ ∃ b, a = f b ∧ q b` will be rewrited as `∃ b, p (f b) ∧ q b`.
+-/
+
+open Lean Meta Qq
+
+namespace existsAndEq
+
+/-- Type for storing chosen branch at `And` nodes. -/
+inductive GoTo
+| left | right
+deriving BEq, Inhabited
+
+/-- Type for storing path in the body expression leading to `a = a'`. We store only chosen
+directions at each `And` node, because there is no branching at `Exists` nodes. -/
+abbrev Path := List GoTo
+
+/-- Qq-fied version of `Expr`. Here we use it to store fvars introduced when unpacking existential
+quantifiers. -/
+abbrev VarQ := (u : Level) × (α : Q(Sort u)) × Q($α)
+
+instance : Inhabited VarQ where
+  default := ⟨default, default, default⟩
+
+/-- Qq-fied version of `Expr` proving some `P : Prop`. -/
+abbrev HypQ := (P : Q(Prop)) × Q($P)
+
+instance : Inhabited HypQ where
+  default := ⟨default, default⟩
+
+/-- Makes `∃ f₁ f₂ ... fₙ, body` where `[f₁, ..., fₙ] = fvars`. -/
+def mkNestedExists (fvars : List VarQ) (body : Expr) : MetaM Expr := do
+  match fvars with
+  | [] => pure body
+  | ⟨_, β, b⟩ :: tl =>
+    let res ← mkNestedExists tl body
+    pure <| ← mkAppOptM ``Exists #[.some β, .some <| ← mkLambdaFVars #[b] res]
```

Comment:
mkAppOptM is probably overkill here, since you have everything you need to build the expression directly